File: src/main/java/Neo4jBenchmark.java
import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import java.util.concurrent.Callable;

/**
 * Main entry point for the Neo4j Benchmark Tool.
 * Parses command-line arguments and delegates tasks to the BenchmarkRunner.
 */
@Command(name = "neo4j-benchmark-tool", mixinStandardHelpOptions = true,
        description = "A benchmark tool for Neo4j.")
public class Neo4jBenchmark implements Callable<Integer> {

    // --- Command Line Arguments ---
    @Option(names = "--uri", description = "Neo4j Bolt URI.", defaultValue = "bolt://localhost:7687")
    private String uri;

    @Option(names = "--user", description = "Neo4j username.", defaultValue = "neo4j")
    private String user;

    @Option(names = "--password", description = "Neo4j password.", defaultValue = "password")
    private String password;

    @Option(names = "--graph-name", required = true, description = "The name of the graph to be tested.")
    private String graphName;

    @Option(names = "--data-path", description = "The path of the data file to be loaded.")
    private String dataPath;

    @Option(names = "--command", required = true, description = "The command to execute: ${COMPLETION-CANDIDATES}")
    private CommandType command;

    @Option(names = "--undirected", description = "Specifies if the graph is undirected.")
    private boolean undirected;

    @Option(names = "--read-ratio", description = "The ratio of read operations in the throughput test.", defaultValue = "0.5")
    private double readRatio;

    @Option(names = "--algm", description = "The GDS algorithm to be tested.", defaultValue = "all")
    private String algm;
    
    // --- REFACTORED OPTION ---
    // This option now specifies the path for the clean, final results file.
    // The shell script is responsible for providing this path.
    @Option(names = "--result-path", required = true, description = "Path to save the clean benchmark result file.")
    private String resultPath;

    /**
     * Main execution logic, called by picocli after parsing arguments.
     * @return 0 on success, 1 on failure.
     */
    @Override
    public Integer call() {
        // Pass the resultPath to the BenchmarkRunner.
        try (BenchmarkRunner benchmarker = new BenchmarkRunner(uri, user, password, graphName, resultPath)) {
            switch (command) {
                case load_graph:
                    benchmarker.dropGraph();
                    benchmarker.createSchema(false);
                    benchmarker.loadData(dataPath, undirected);
                    break;
                case load_property:
                    benchmarker.dropGraph();
                    benchmarker.createSchema(true);
                    benchmarker.loadPropertyGraph(dataPath);
                    break;
                case test_basic_op:
                    benchmarker.basicOpTest();
                    break;
                case test_read_write_op:
                    benchmarker.readWriteTest(readRatio);
                    break;
                case test_property:
                    benchmarker.propertyOpTest();
                    break;
                case test_algm:
                    benchmarker.testAlgm(algm);
                    break;
                default:
                    System.err.println("Command not implemented: " + command);
                    return 1;
            }
        } catch (Exception e) {
            // Print stack trace to stderr, which will be captured by the log file.
            e.printStackTrace();
            return 1;
        }
        return 0;
    }

    public static void main(String[] args) {
        int exitCode = new CommandLine(new Neo4jBenchmark()).execute(args);
        System.exit(exitCode);
    }
}

File: src/neo4j_benchmark.py
# file: neo4j_benchmark.py

import typer
from enum import Enum
from benchmark_runner import BenchmarkRunner

app = typer.Typer(
    name="neo4j-benchmark",
    help="A benchmark tool for Neo4j, mirroring the Nebula Graph test suite.",
    add_completion=False
)

class CommandType(str, Enum):
    load_graph = "load_graph"
    # load_property = "load_property"
    test_basic_op = "test_basic_op"
    test_read_write_op = "test_read_write_op"
    # test_property = "test_property"
    test_algm = "test_algm"

@app.command()
def run(
    command: CommandType = typer.Option(..., "--command", help="The command to execute.", case_sensitive=False),
    graph_name: str = typer.Option(..., "--graph-name", help="The name of the graph to be tested."),
    report_path: str = typer.Option(None, "--report-path", help="Path to save the detailed benchmark report file."),
    uri: str = typer.Option("bolt://localhost:7687", "--uri", help="Neo4j Bolt URI."),
    user: str = typer.Option("neo4j", "--user", help="Neo4j username."),
    password: str = typer.Option("password", "--password", help="Neo4j password."),
    data_path: str = typer.Option(None, "--data-path", help="The path of the data file to be loaded."),
    undirected: bool = typer.Option(False, "--undirected", help="Whether the graph is undirected."),
    read_ratio: float = typer.Option(0.5, "--read-ratio", help="The ratio of read/lookup operations."),
    algm: str = typer.Option("all", "--algm", help="The algorithm to be tested.")
):
    """
    Run a specific benchmark command for Neo4j.
    """
    try:
        with BenchmarkRunner(uri, user, password, graph_name, report_path) as benchmarker:
            if command == CommandType.load_graph:
                if not data_path:
                    typer.echo("Error: --data-path is required for --command load_graph", err=True)
                    raise typer.Exit(code=1)
                benchmarker.drop_graph()
                benchmarker.create_schema(is_property_graph=False)
                benchmarker.load_data(data_path, undirected)

            elif command == CommandType.test_basic_op:
                benchmarker.basic_op_test()

            elif command == CommandType.test_read_write_op:
                benchmarker.read_write_test(read_ratio)
                
            elif command == CommandType.test_algm:
                benchmarker.test_algm(algm)

    except Exception as e:
        typer.echo(f"\nAn error occurred: {e}", err=True)
        # import traceback
        # traceback.print_exc() # Uncomment for detailed debugging
        raise typer.Exit(code=1)
    
if __name__ == "__main__":
    app()